src/prefect/cli/cloud.py

```diff
@@ -150,13 +299,81 @@ async def login(
     Creates a new profile configured to use the specified PREFECT_API_KEY.
     Uses a previously configured profile if it exists.
     """
+    if not is_interactive() and (not key or not workspace_handle):
+        exit_with_error(
+            "When not using an interactive terminal, you must supply a `--key` and `--workspace`."
+        )
+
+    profiles = load_profiles()
+    current_profile = get_settings_context().profile
+
+    if key and PREFECT_API_KEY.value() == key:
+        exit_with_success("This profile is already authenticated with that key.")
+
+    already_logged_in_profiles = []
+    for name, profile in profiles.items():
+        profile_key = profile.settings.get(PREFECT_API_KEY)
+        if (
+            not key
+            and profile_key is not None
+            or (key and profile_key == key)
+            and await check_key_is_valid_for_login(profile_key)
+        ):
```

Comment:
This was kinda rough to reason through on first glance. 

```suggestion
        only_profile_key_provided = not key and profile_key is not None
        key_matches_profile_key = key and profile_key == key
        if (
            (only_profile_key_provided or key_matches_profile_key)
            and await check_key_is_valid_for_login(profile_key)
        ):
```

Reply:
I'm also in general wondering if a few of these shouldn't be moved out to separate functions just to keep login a little more manageable to work through. It's ~150 lines long at the moment.

`already_logged_in_profiles = get_logged_in_profiles(key, profiles)`, etc.

Reply:
I attempted to address this in https://github.com/PrefectHQ/prefect/pull/7334/commits/244d7269a341db403b9986e58db2d712d7024563

Yeah this sort of thing should be probably be abstracted out into a different function. This pull request is in the stage where I have not gone through and moved things into utilities, basically in an attempt to get all the logic in one place first. I think a later refactor would be prudent.